### PR TITLE
tailcfg,ipn/localapi,client/tailscale: add QueryFeature endpoint

### PR DESCRIPTION
**Pull request by soniaappasamy**
Updates tailscale/corp#10577

### CODE REVIEW

File: client/tailscale/localclient.go [hunk 1]
```diff
@@ -1124,6 +1124,25 @@ func (lc *LocalClient) DeleteProfile(ctx context.Context, profile ipn.ProfileID)
 	return err
 }
 
+// EnableFeature makes a request for instructions on how to enable a
```
Comment (bradfitz) [inline client/tailscale/localclient.go:1127]: Kinda a weird name in that it doesn't actually enable it.

It's more like "QueryFeatureAvailability" or something, which is kinda a mouthful. Maybe "QueryFeature" or "CheckFeature"?

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -2257,6 +2257,46 @@ type SSHRecordingAttempt struct {
...
+// to enable a feature, such as Funnel, for the node's tailnet.
+//
+// This request itself DOES NOT modify tailnet configuration to enable the
+// feature, but returns information that can be presented to the user about
+// where/how to enable the feature. See EnableFeatureResponse for response
+// structure.
+type EnableFeatureRequest struct {
+	// NodeKey is the client's current node key.
+	NodeKey key.NodePublic `json:",omitempty"`
+}
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:2270]: now that I see this written out, it's kinda weird that the feature string isn't in the request but lives outside in the URL.

I don't think REST-y purity should be our goal here: this is a weirdo API over noise only that end users can't really hit.  So the server endpoint could just be `/machine/feature` or something?

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -2257,6 +2257,46 @@ type SSHRecordingAttempt struct {
...
+// EnableFeatureResponse is the response to an EnableFeatureRequest.
+type EnableFeatureResponse struct {
+	// Complete is true when the feature is already enabled.
+	Complete bool `json:",omitempty"`
+
+	// Text holds lines to display in the CLI with information
+	// about the feature and how to enable it.
+	//
+	// Lines are separated by newline characters. No trailing
+	// newline.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:2281]: last sentence isn't a full sentence. But maybe leave more wiggle worm. "The final newline may be omitted." ?

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -2257,6 +2257,46 @@ type SSHRecordingAttempt struct {
...
+	Text string `json:",omitempty"`
+
+	// URL is filled with the link for the user to visit to take
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:2284]: can drop the passive voice. "URL is the link for the user to visit to..."

File: client/tailscale/localclient.go (hunk 1 above)
Comment (soniaappasamy) [reply client/tailscale/localclient.go:1127]: Yea I was going back and forth on naming too - I like `QueryFeature`, was also considering `RequestFeature` but then that's a lot of "request request" repetition

File: tailcfg/tailcfg.go (hunk 2 above)
Comment (soniaappasamy) [reply tailcfg/tailcfg.go:2270]: Yeah - can do `/machine/feature` and move feature to the request

File: client/tailscale/localclient.go (hunk 1 above)
Comment (soniaappasamy) [reply client/tailscale/localclient.go]: Okay, updated to `QueryFeature`. I'm still not totally in love with the name, but can start here and update if I come up with a better one.

File: client/tailscale/localclient.go [hunk 5]
```diff
@@ -1124,6 +1124,25 @@ func (lc *LocalClient) DeleteProfile(ctx context.Context, profile ipn.ProfileID)
 	return err
 }
 
+// QueryFeature makes a request for instructions on how to enable a
+// feature, such as Funnel, for the node's tailnet.
+//
+// This request itself DOES NOT directly enable the feature on behalf
```
Comment (bradfitz) [inline client/tailscale/localclient.go:1130]: I think this shouting is probably no longer necessary given the new name of the method.

File: client/tailscale/localclient.go [hunk 6]
```diff
@@ -1124,6 +1124,25 @@ func (lc *LocalClient) DeleteProfile(ctx context.Context, profile ipn.ProfileID)
...
+// of the node, but rather returns information that can be presented
+// to the acting user about where/how to enable the feature.
+//
+// If relevant, this includes a control URL the user can visit to
+// explicitly consent to using the feature. LocalClient.WatchIPNBus
+// can be used to block on the feature being enabled.
+func (lc *LocalClient) QueryFeature(ctx context.Context, feature string) (*tailcfg.QueryFeatureResponse, error) {
```
Comment (bradfitz) [inline client/tailscale/localclient.go:1137]: any doc guidance on what strings this feature might be?

File: ipn/localapi/localapi.go [hunk 7]
```diff
@@ -1932,6 +1933,64 @@ func (h *Handler) serveProfiles(w http.ResponseWriter, r *http.Request) {
...
+	b, err := json.Marshal(&tailcfg.QueryFeatureRequest{
+		NodeKey: nm.NodeKey,
+		Feature: feature,
+	})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	req, err := http.NewRequest("POST", "https://unused/machine/feature/enable", bytes.NewReader(b))
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:1975]: NewRequestWithContext(r.Context(), "POST" ...

File: ipn/localapi/localapi.go [hunk 8]
```diff
@@ -1932,6 +1933,64 @@ func (h *Handler) serveProfiles(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	resp, err := h.b.DoNoiseRequest(req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+	w.WriteHeader(resp.StatusCode)
```
Comment (bradfitz) [inline ipn/localapi/localapi.go:1987]: you want to copy the Content-Type header at least here before the WriteHeader?

File: tailcfg/tailcfg.go [hunk 9]
```diff
@@ -2257,6 +2257,48 @@ type SSHRecordingAttempt struct {
 	FailureMessage string
 }
 
+// QueryFeatureRequest is a request sent to "/machine/feature/enable"
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:2260]: do you want "enable" here? not "query" now?

**Review by bradfitz — APPROVED** [review]

File: client/tailscale/localclient.go (hunk 6 above)
Comment (soniaappasamy) [reply client/tailscale/localclient.go:1139]: added

**Pull request merged**